subt/cloudsim2osgar.py

```diff
@@ -145,41 +145,47 @@ def __init__(self, robot_name, robot_config, robot_is_marsupial):
         else:
             topics.append(('/' + robot_name + '/imu/data', Imu, self.imu, ('rot', 'acc', 'orientation')))
 
-        outputs = functools.reduce(operator.add, (t[-1] for t in topics)) + ('origin',)
+        outputs = functools.reduce(operator.add, (t[-1] for t in topics)) + ('origin', 'robot_name',)
         self.bus.register(outputs)
 
         # origin
         self.origin = None
-        try:
-            ORIGIN_SERVICE_NAME = "/subt/pose_from_artifact_origin"
-            rospy.wait_for_service(ORIGIN_SERVICE_NAME)
-            origin_service = rospy.ServiceProxy(ORIGIN_SERVICE_NAME, PoseFromArtifact)
-            origin_request = String()
-            origin_request.data = robot_name
-            origin_response = origin_service(origin_request)
-            if origin_response.success:
-                origin_pose = origin_response.pose.pose
-                origin_position = origin_response.pose.pose.position
-                origin_orientation = origin_response.pose.pose.orientation
-                self.origin = (
-                        (origin_position.x, origin_position.y, origin_position.z),
-                        (origin_orientation.x, origin_orientation.y, origin_orientation.z,
-                            origin_orientation.w))
-        except rospy.ServiceException:
-            rospy.logerror("Failed to get origin.")
-        if self.origin is None:
-            # Lack of ccordinates signals an error.
-            self.bus.publish('origin', (robot_name,))
-        else:
-            origin_msg = (robot_name,) + self.origin[0] + self.origin[1]
-            self.bus.publish('origin', origin_msg)
+        origin_retry_delay = 0.2
+        ORIGIN_RETRY_EXPONENTIAL_BACKOFF = 1.3
+        MAX_ORIGIN_RETRY_DELAY = 2.0
+        while self.origin is None:
+            try:
+                ORIGIN_SERVICE_NAME = "/subt/pose_from_artifact_origin"
+                rospy.wait_for_service(ORIGIN_SERVICE_NAME)
+                origin_service = rospy.ServiceProxy(ORIGIN_SERVICE_NAME, PoseFromArtifact)
+                origin_request = String()
+                origin_request.data = robot_name
```

Comment:
Would you mind moving all the lines between `try` and this before the `while` loop? They should be created only once.

Reply:
OK, good point - fixed.